ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionCommitter.scala

```diff
@@ -387,9 +388,27 @@ private[kvutils] class TransactionCommitter(
     )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
       val transaction = transactionEntry.submission.getTransaction
       val nodes = transaction.getNodesList.asScala
-      val nodesToKeep = nodes.iterator.collect {
-        case node if node.hasCreate || node.hasExercise => node.getNodeId
-      }.toSet
+      val nodeMap: Map[String, TransactionOuterClass.Node] =
+        nodes.view.map(n => n.getNodeId -> n).toMap
+
+      @tailrec
+      def goNodesToKeep(todo: List[String], result: Set[String]): Set[String] = todo match {
+        case Nil => result
+        case head :: tail =>
+          import TransactionOuterClass.Node.NodeTypeCase
+          val node = nodeMap.get(head).getOrElse(throw Err.InvalidSubmission(s"Invalid transaction node id $head"))
```

Comment:
If there is a more appropriate error I should throw here let me know. Encoding/decoding didn’t really seem appropriate. InternalError also would work I guess.